storage/innobase/trx/trx0purge.cc

```diff
@@ -163,6 +85,15 @@ purge_graph_build()
 	return(fork);
 }
 
+/* Define it non-inline as pushing to priority queue is rather complex
+operation */
+void purge_sys_t::enqueue(trx_rseg_t &rseg)
+{
+  ut_ad(srv_operation == SRV_OPERATION_RESTORE ||
+        mysql_mutex_is_owner(&pq_mutex));
```

Comment:
We could add explicit
```c++
ATTRIBUTE_NOINLINE /* this is a rather complex operation */
```
I am a bit curious about this relaxation. Why do we need it during `mariadb-backup --prepare`? If the relaxation cannot be avoided, the assertion would have to be enclosed within
```c
#ifdef SAFE_MUTEX
```
There is no such issue with `mysql_mutex_assert_owner`.

Reply:
The relaxation is due to the following code:

```
dberr_t trx_lists_init_at_db_start()                                            
{                                                                               
...                                                                             
        if (srv_operation == SRV_OPERATION_RESTORE) {                           
                /* mariabackup --prepare only deals with                        
                the redo log and the data files, not with                       
                transactions or the data dictionary. */                         
                return trx_rseg_array_init();                                   
        }                                                                       
                                                                                
        if (srv_force_recovery >= SRV_FORCE_NO_UNDO_LOG_SCAN) {                 
                return DB_SUCCESS;                                              
        }                                                                       
                                                                                
        purge_sys.create();                                                     
        dberr_t err = trx_rseg_array_init();                                    
...                                                                             
}
```

Reply:
Oh, the relaxation is not neccessary because `trx_rseg_mem_restore()` quits before initializing undo log lists.

Reply:
I see that `ATTRIBUTE_NOINLINE` is used in header files. What transaction unit(s) contain(s) the compiled function's code? All units, which include the corresponding header file?